### ISSUE TITLE
[RFC] Adding S3 support

### ISSUE THREAD

**Issue by rande** [spf13/afero#11]
Hello,

Do you have any plan to support S3 bucket ? or other remote FS ?

**Comment by spf13** [comment]
That would be a great thing to contribute. This is primarily a framework but it would be great if others would participate and add support for remote FS.

**Comment by chonthu** [comment]
I could see an interface that ties in https://github.com/mitchellh/goamz/tree/master/s3

**Comment by spf13** [comment]
please contribute it. Not sure if that is the best library anymore now that Amazon has decided to officially support golang.
https://github.com/aws/aws-sdk-go/tree/master/service/s3

**Comment by chonthu** [comment]
Seems a bit bloated, Ill take a look. But ideally the package would work like below?

```
package afero

import (
	"github.com/mitchellh/goamz/aws"
	"github.com/mitchellh/goamz/s3"
	"testing"
)

func TestS3OpenFile(t *testing.T) {

	auth, err := aws.EnvAuth()
	if err != nil {
		t.Fatal(err)
	}

	client := s3.New(auth, aws.USEast)
	bucket, err := client.Bucket("bucket_name")
	if err != nil {
		t.Fatal(err)
	}

	var AppFs Fs = S3Fs{
		Bucket: bucket,
	}

	AppFs.Open("files/+131217-best-country-10-justin-moore-150x150.jpg")
}
```

**Comment by spf13** [comment]
I like that syntax and I think it makes a lot of sense to have the Auth portion outside of Afero and make Afero just operate on the bucket. 

If contributing, I'm happy to have it in the main repo if you want, but feel it should be an optional backend so it should go in a subdirectory (and package) rather than the main one as it introduces a large new dependency. Does that make sense?

**Comment by chonthu** [comment]
Agreed 100%, let me look into the official aws package tonight and see what makes sense for a PR

**Comment by rande** [comment]
I think Afero current interface cannot work properly with S3 as there is no directory support in S3. You can of course simulate a directory structure but still...

It can be great if afero can have 2 interface, one for interacting with binary content (k/v store) like s3 or any object storage from any cloud providers and another one dealing with directory management. For now I have removed afero from my side project in favor a custom one: https://github.com/rande/gonode/blob/master/vault/vault.go#L31-L36

But I will more than happy to remove my code, if I found a more stable and supported library like afero.

**Comment by spf13** [comment]
All of the similar libraries emulate a directory structure and treat s3
like a filesystem. It's a well established approach.

**Comment by chonthu** [comment]
I also agree with @spf13 here, seems like common practice @rande. Started on the sub dir @spf13 

https://github.com/spf13/afero/compare/master...chonthu:feature/s3fs?expand=1

Starting to tackle each method next

**Comment by thoas** [comment]
@spf13 I have made a similar abstraction for FS, available [here](https://github.com/thoas/gostorages/blob/master/s3.go), I can contribute it back to afero or @chonthu

**Comment by raff** [comment]
I think one thing that would really be helpful to make creating more back-ends easier is to clearly specify what the interface methods should do (i.e. does a Create file return an error if the file already exists or override ? Or there should be a way to do both ?). For some backend it's easier to do one way but not the other and there should be a clear way of specify that and/or return the appropriate errors.

Also, again, to facilitate adding more backend, I think it would be useful to have a "read-only" mode, or an "append only" mode (I guess this would be a "property" of a backend, versus a user settable mode).

I was thinking of adding a "zip" backend (I have a project where I ended up implementing a wrapper for "zip", "s3" and such, that would benefit from switching to "afero", and I can probably reuse some of my code). And, for example, for "zip" I can easily see a "zip read-only" backend (for extracting resources from an existing file) and a "zip" backend where new files are added to the zip file. And I just looked at the archive/zip package and you can only open a zip file for "reading" or for "writing" (and when writing only append operations are supported). Of course it would be possible to implement the full interface by copying and rewriting a zip file, but I don't think it's worth the effort (and in the few cases where a user really need that they could easily implement this on top of the provided implementation).

**Comment by chonthu** [comment]
Hey @thoas I could always use the help and suggestions! Here is my latest thoughts: https://github.com/spf13/afero/compare/master...chonthu:feature/s3fs?expand=1

@raff I agree, will try incorporate that into my s3 flavor. Would love to see how you go about your zip implementation.

**Comment by mbertschler** [comment]
@raff As far as I understand it is to provide an interface that is a drop in replacement for the standard os package functions. For example the doc for os.Create states:
> Create creates the named file with mode 0666 (before umask), truncating it if it already exists.

I just added a test to check for this behavior.

A great way to test new backends is to add it to this line at the top of the `afero_test.go` file, as all the tests loop over this array and test if they behave the same.

```go
var Fss = []Fs{&MemMapFs{}, &OsFs{}}
```

**Comment by pgruenbacher** [comment]
If implement something similar to afero-s3 or afero-gcs, my thinking would be to have the interface embedding a database layer to manage the folders and tracking of the cloud-storage meta information.

**Comment by fgm** [comment]
3 years later, did this go anywhere ? There seems to have been a lot of interest at the time, but none since. Am I missing something ? Maybe the fact that [S3 can now be accessed over SFTP](https://aws.amazon.com/sftp/?nc1=f_ls) solves the issue ? Or maybe using a FUSE S3FS implementation is now a general solution ?

* #73 made other suggestions, but appear to have no suggested code.
* https://github.com/ulule/gostorages has its own implementation, but nothing directly compatible 
* https://github.com/rande/gonode/blob/1590ae4e56588b06649d9b3b35fca92d12b3f1c8/core/vault/vault_driver_s3.go also has an incompatible implementation
* https://github.com/chonthu/afero-s3 exists, but the [Stable ?](https://github.com/chonthu/aferoS3/issues/1) issue on it remained unanswered since 2016

**Comment by JessicaCGlenn** [comment]
Hey, I hit this as well, so I forked @chonthu's code from 4 years ago and have started bringing it back to life a little bit:

https://github.com/qugit/aferos3

I still have to add some more tests and documentation, and I think there are definitely some ways that I could make the current code more robust, so contributions are definitely welcome!

**Comment by chonthu** [comment]
Thanks, @JessicaCGlenn, created a feature branch with the new sdk here. Happy to accept PRs. I'll start a minio support branch as well

https://github.com/chonthu/aferoS3/tree/feature/aws-sdk

**Comment by chonthu** [comment]
It now merged: https://github.com/chonthu/aferoS3

**Comment by kargakis** [comment]
Last PR on this: https://github.com/spf13/afero/pull/90
